In [3]:
import pandas as pd

In [4]:
data = pd.read_csv('../M1_final.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28820 entries, 0 to 28819
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   MONTH              28820 non-null  int64  
 1   DAY_OF_MONTH       28820 non-null  int64  
 2   DAY_OF_WEEK        28820 non-null  int64  
 3   OP_UNIQUE_CARRIER  28820 non-null  object 
 4   TAIL_NUM           28820 non-null  object 
 5   DEST               28820 non-null  object 
 6   DEP_DELAY          28820 non-null  int64  
 7   CRS_ELAPSED_TIME   28820 non-null  int64  
 8   DISTANCE           28820 non-null  int64  
 9   CRS_DEP_M          28820 non-null  int64  
 10  DEP_TIME_M         28820 non-null  int64  
 11  CRS_ARR_M          28820 non-null  int64  
 12  Temperature        28820 non-null  int64  
 13  Dew Point          28820 non-null  object 
 14  Humidity           28820 non-null  int64  
 15  Wind               28818 non-null  object 
 16  Wind Speed         288

# Dataset Analysis for Research Paper

This notebook analyzes the M1_final.csv dataset to extract information needed for the dataset section of the research paper.

## Dataset Section Checklist:
1. Dataset name and source
2. Data collection details (if applicable)
3. Number of samples
4. Number of features
5. Description of each feature
6. Description of target variable
7. Class distribution (just raw numbers)
8. Missing/duplicate records (brief)
9. Time period of data (if applicable)
10. Sampling rate (if time-series)
11. Basic preprocessing (minimal)
12. Ethical/privacy notes (if needed)

In [5]:
# Basic Dataset Information
print("="*60)
print("DATASET OVERVIEW")
print("="*60)
print(f"Dataset Name: M1_final.csv")
print(f"Number of samples (rows): {data.shape[0]:,}")
print(f"Number of features (columns): {data.shape[1]:,}")
print(f"Dataset shape: {data.shape}")
print(f"Memory usage: {data.memory_usage(deep=True).sum() / 1024**2:.2f} MB")
print("="*60)

DATASET OVERVIEW
Dataset Name: M1_final.csv
Number of samples (rows): 28,820
Number of features (columns): 23
Dataset shape: (28820, 23)
Memory usage: 12.55 MB


In [7]:
# Feature Analysis
print("\n" + "="*60)
print("FEATURE DESCRIPTIONS")
print("="*60)

# Display column names and data types
print("\nColumns and Data Types:")
for i, (col, dtype) in enumerate(zip(data.columns, data.dtypes), 1):
    non_null_count = data[col].count()
    print(f"{i:2d}. {col:<20} | {str(dtype):<10} | Non-null: {non_null_count:,}")

# Let's examine the first few rows to understand the data better
print("\n" + "="*60)
print("SAMPLE DATA (First 5 rows)")
print("="*60)
data.head()


FEATURE DESCRIPTIONS

Columns and Data Types:
 1. MONTH                | int64      | Non-null: 28,820
 2. DAY_OF_MONTH         | int64      | Non-null: 28,820
 3. DAY_OF_WEEK          | int64      | Non-null: 28,820
 4. OP_UNIQUE_CARRIER    | object     | Non-null: 28,820
 5. TAIL_NUM             | object     | Non-null: 28,820
 6. DEST                 | object     | Non-null: 28,820
 7. DEP_DELAY            | int64      | Non-null: 28,820
 8. CRS_ELAPSED_TIME     | int64      | Non-null: 28,820
 9. DISTANCE             | int64      | Non-null: 28,820
10. CRS_DEP_M            | int64      | Non-null: 28,820
11. DEP_TIME_M           | int64      | Non-null: 28,820
12. CRS_ARR_M            | int64      | Non-null: 28,820
13. Temperature          | int64      | Non-null: 28,820
14. Dew Point            | object     | Non-null: 28,820
15. Humidity             | int64      | Non-null: 28,820
16. Wind                 | object     | Non-null: 28,818
17. Wind Speed           | int64      | N

,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,OP_UNIQUE_CARRIER,TAIL_NUM,DEST,DEP_DELAY,CRS_ELAPSED_TIME,DISTANCE,CRS_DEP_M,...,Dew Point,Humidity,Wind,Wind Speed,Wind Gust,Pressure,Condition,sch_dep,sch_arr,TAXI_OUT
0,11,1,5,B6,N828JB,CHS,-1,124,636,324,...,34,58,W,25,38,29.86,Fair / Windy,9,17,14
1,11,1,5,B6,N992JB,LAX,-7,371,2475,340,...,34,58,W,25,38,29.86,Fair / Windy,9,17,15
2,11,1,5,B6,N959JB,FLL,40,181,1069,301,...,34,58,W,25,38,29.86,Fair / Windy,9,17,22
3,11,1,5,B6,N999JQ,MCO,-2,168,944,345,...,34,58,W,25,38,29.86,Fair / Windy,9,17,12
4,11,1,5,DL,N880DN,ATL,-4,139,760,360,...,32,58,W,24,35,29.91,Fair / Windy,9,17,13


In [8]:
# Identify Target Variable and Feature Categories
print("\n" + "="*60)
print("TARGET VARIABLE IDENTIFICATION")
print("="*60)

# Based on the columns, this appears to be a flight dataset
# Let's identify the likely target variable (probably DEP_DELAY or TAXI_OUT)
potential_targets = ['DEP_DELAY', 'TAXI_OUT']

for target in potential_targets:
    if target in data.columns:
        print(f"\nPotential Target Variable: {target}")
        print(f"Description: {target}")
        print(f"Data type: {data[target].dtype}")
        print(f"Unique values: {data[target].nunique()}")
        print(f"Range: {data[target].min()} to {data[target].max()}")
        print(f"Mean: {data[target].mean():.2f}")
        print(f"Standard Deviation: {data[target].std():.2f}")

# Let's also check the distribution of DEP_DELAY (likely the main target)
if 'DEP_DELAY' in data.columns:
    print(f"\nDEP_DELAY Distribution:")
    print(f"Negative delays (early departures): {(data['DEP_DELAY'] < 0).sum():,}")
    print(f"On-time departures (0 delay): {(data['DEP_DELAY'] == 0).sum():,}")
    print(f"Positive delays (late departures): {(data['DEP_DELAY'] > 0).sum():,}")
    
    # Create delay categories for classification analysis
    data_temp = data.copy()
    data_temp['DELAY_CATEGORY'] = pd.cut(data_temp['DEP_DELAY'], 
                                        bins=[-float('inf'), 0, 15, 60, float('inf')],
                                        labels=['Early/OnTime', 'Short_Delay', 'Medium_Delay', 'Long_Delay'])
    print(f"\nDelay Categories Distribution:")
    print(data_temp['DELAY_CATEGORY'].value_counts())


TARGET VARIABLE IDENTIFICATION

Potential Target Variable: DEP_DELAY
Description: DEP_DELAY
Data type: int64
Unique values: 342
Range: -22 to 1276
Mean: 6.37
Standard Deviation: 38.74

Potential Target Variable: TAXI_OUT
Description: TAXI_OUT
Data type: int64
Unique values: 37
Range: 5 to 41
Mean: 20.86
Standard Deviation: 6.85

DEP_DELAY Distribution:
Negative delays (early departures): 19,549
On-time departures (0 delay): 1,466
Positive delays (late departures): 7,805

Delay Categories Distribution:
DELAY_CATEGORY
Early/OnTime    21015
Short_Delay      3931
Medium_Delay     2556
Long_Delay       1318
Name: count, dtype: int64


In [9]:
# Missing Values and Data Quality Analysis
print("\n" + "="*60)
print("DATA QUALITY ANALYSIS")
print("="*60)

# Missing values
print("Missing Values per Column:")
missing_values = data.isnull().sum()
missing_percentage = (missing_values / len(data)) * 100

for col in data.columns:
    if missing_values[col] > 0:
        print(f"{col:<20}: {missing_values[col]:,} ({missing_percentage[col]:.2f}%)")

if missing_values.sum() == 0:
    print("No missing values found in the dataset!")
else:
    print(f"\nTotal missing values: {missing_values.sum():,}")

# Check for duplicate records
duplicates = data.duplicated().sum()
print(f"\nDuplicate Records: {duplicates:,}")

# Data completeness
completeness = ((data.count().sum()) / (data.shape[0] * data.shape[1])) * 100
print(f"Data Completeness: {completeness:.2f}%")


DATA QUALITY ANALYSIS
Missing Values per Column:
Wind                : 2 (0.01%)

Total missing values: 2

Duplicate Records: 0
Data Completeness: 100.00%


In [10]:
# Feature Categories and Detailed Descriptions
print("\n" + "="*60)
print("DETAILED FEATURE DESCRIPTIONS")
print("="*60)

# Categorize features
temporal_features = ['MONTH', 'DAY_OF_MONTH', 'DAY_OF_WEEK', 'CRS_DEP_M', 'DEP_TIME_M', 'CRS_ARR_M', 'sch_dep', 'sch_arr']
flight_features = ['OP_UNIQUE_CARRIER', 'TAIL_NUM', 'DEST', 'CRS_ELAPSED_TIME', 'DISTANCE']
weather_features = ['Temperature', 'Dew Point', 'Humidity', 'Wind', 'Wind Speed', 'Wind Gust', 'Pressure', 'Condition']
operational_features = ['DEP_DELAY', 'TAXI_OUT']

print("Feature Categories:")
print(f"1. Temporal Features ({len(temporal_features)}): {', '.join(temporal_features)}")
print(f"2. Flight Features ({len(flight_features)}): {', '.join(flight_features)}")
print(f"3. Weather Features ({len(weather_features)}): {', '.join(weather_features)}")
print(f"4. Operational Features ({len(operational_features)}): {', '.join(operational_features)}")

print(f"\nTotal Features: {len(temporal_features) + len(flight_features) + len(weather_features) + len(operational_features)}")

# Detailed descriptions for research paper
feature_descriptions = {
    'MONTH': 'Month of the flight (1-12)',
    'DAY_OF_MONTH': 'Day of the month (1-31)',
    'DAY_OF_WEEK': 'Day of the week (1-7)',
    'OP_UNIQUE_CARRIER': 'Airline carrier code',
    'TAIL_NUM': 'Aircraft tail number identifier',
    'DEST': 'Destination airport code',
    'DEP_DELAY': 'Departure delay in minutes (negative = early, 0 = on-time, positive = delayed)',
    'CRS_ELAPSED_TIME': 'Scheduled elapsed flight time in minutes',
    'DISTANCE': 'Flight distance in miles',
    'CRS_DEP_M': 'Scheduled departure time in minutes from midnight',
    'DEP_TIME_M': 'Actual departure time in minutes from midnight',
    'CRS_ARR_M': 'Scheduled arrival time in minutes from midnight',
    'Temperature': 'Ambient temperature',
    'Dew Point': 'Dew point temperature',
    'Humidity': 'Relative humidity percentage',
    'Wind': 'Wind direction and characteristics',
    'Wind Speed': 'Wind speed',
    'Wind Gust': 'Wind gust speed',
    'Pressure': 'Atmospheric pressure',
    'Condition': 'Weather condition description',
    'sch_dep': 'Scheduled departure time (encoded)',
    'sch_arr': 'Scheduled arrival time (encoded)',
    'TAXI_OUT': 'Taxi-out time in minutes'
}

print(f"\n{'Feature':<20} | {'Description'}")
print("-" * 80)
for feature, description in feature_descriptions.items():
    print(f"{feature:<20} | {description}")


DETAILED FEATURE DESCRIPTIONS
Feature Categories:
1. Temporal Features (8): MONTH, DAY_OF_MONTH, DAY_OF_WEEK, CRS_DEP_M, DEP_TIME_M, CRS_ARR_M, sch_dep, sch_arr
2. Flight Features (5): OP_UNIQUE_CARRIER, TAIL_NUM, DEST, CRS_ELAPSED_TIME, DISTANCE
3. Weather Features (8): Temperature, Dew Point, Humidity, Wind, Wind Speed, Wind Gust, Pressure, Condition
4. Operational Features (2): DEP_DELAY, TAXI_OUT

Total Features: 23

Feature              | Description
--------------------------------------------------------------------------------
MONTH                | Month of the flight (1-12)
DAY_OF_MONTH         | Day of the month (1-31)
DAY_OF_WEEK          | Day of the week (1-7)
OP_UNIQUE_CARRIER    | Airline carrier code
TAIL_NUM             | Aircraft tail number identifier
DEST                 | Destination airport code
DEP_DELAY            | Departure delay in minutes (negative = early, 0 = on-time, positive = delayed)
CRS_ELAPSED_TIME     | Scheduled elapsed flight time in minutes
DIS

In [11]:
# Temporal Analysis
print("\n" + "="*60)
print("TEMPORAL ANALYSIS")
print("="*60)

# Analyze temporal patterns
print("Month Distribution:")
month_dist = data['MONTH'].value_counts().sort_index()
for month, count in month_dist.items():
    print(f"Month {month:2d}: {count:,} flights ({count/len(data)*100:.1f}%)")

print(f"\nDay of Week Distribution:")
dow_dist = data['DAY_OF_WEEK'].value_counts().sort_index()
for dow, count in dow_dist.items():
    day_names = {1: 'Monday', 2: 'Tuesday', 3: 'Wednesday', 4: 'Thursday', 
                 5: 'Friday', 6: 'Saturday', 7: 'Sunday'}
    day_name = day_names.get(dow, f'Day {dow}')
    print(f"{day_name}: {count:,} flights ({count/len(data)*100:.1f}%)")

# Time period estimation
print(f"\nTime Period Analysis:")
print(f"Data spans across {data['MONTH'].nunique()} unique months")
print(f"Data covers {data['DAY_OF_MONTH'].nunique()} unique days of month")
print(f"Data includes all {data['DAY_OF_WEEK'].nunique()} days of the week")


TEMPORAL ANALYSIS
Month Distribution:
Month  1: 9,869 flights (34.2%)
Month 11: 9,769 flights (33.9%)
Month 12: 9,182 flights (31.9%)

Day of Week Distribution:
Monday: 4,051 flights (14.1%)
Tuesday: 4,065 flights (14.1%)
Wednesday: 4,013 flights (13.9%)
Thursday: 4,135 flights (14.3%)
Friday: 4,708 flights (16.3%)
Saturday: 3,698 flights (12.8%)
Sunday: 4,150 flights (14.4%)

Time Period Analysis:
Data spans across 3 unique months
Data covers 31 unique days of month
Data includes all 7 days of the week


In [12]:
# Categorical Variables Analysis
print("\n" + "="*60)
print("CATEGORICAL VARIABLES ANALYSIS")
print("="*60)

categorical_cols = data.select_dtypes(include=['object']).columns
print(f"Categorical Variables: {list(categorical_cols)}")

for col in categorical_cols:
    unique_count = data[col].nunique()
    print(f"\n{col}:")
    print(f"  Unique values: {unique_count}")
    if unique_count <= 10:  # Show distribution for variables with few categories
        value_counts = data[col].value_counts()
        for value, count in value_counts.head(10).items():
            print(f"    {value}: {count:,} ({count/len(data)*100:.1f}%)")
    else:  # Just show top categories for variables with many categories
        value_counts = data[col].value_counts()
        print(f"  Top 5 categories:")
        for value, count in value_counts.head(5).items():
            print(f"    {value}: {count:,} ({count/len(data)*100:.1f}%)")

# Weather conditions analysis
if 'Condition' in data.columns:
    print(f"\nWeather Conditions Summary:")
    weather_counts = data['Condition'].value_counts()
    print(f"Most common weather: {weather_counts.index[0]} ({weather_counts.iloc[0]:,} flights)")
    print(f"Weather variety: {len(weather_counts)} different conditions")


CATEGORICAL VARIABLES ANALYSIS
Categorical Variables: ['OP_UNIQUE_CARRIER', 'TAIL_NUM', 'DEST', 'Dew Point', 'Wind', 'Condition']

OP_UNIQUE_CARRIER:
  Unique values: 9
    B6: 9,589 (33.3%)
    DL: 7,680 (26.6%)
    9E: 4,485 (15.6%)
    AA: 3,595 (12.5%)
    AS: 1,111 (3.9%)
    MQ: 1,045 (3.6%)
    YX: 914 (3.2%)
    OO: 311 (1.1%)
    HA: 90 (0.3%)

TAIL_NUM:
  Unique values: 2092
  Top 5 categories:
    N102NN: 108 (0.4%)
    N103NN: 103 (0.4%)
    N2002J: 101 (0.4%)
    N2016J: 100 (0.3%)
    N115NN: 99 (0.3%)

DEST:
  Unique values: 65
  Top 5 categories:
    LAX: 2,974 (10.3%)
    SFO: 1,678 (5.8%)
    BOS: 1,243 (4.3%)
    FLL: 1,008 (3.5%)
    MCO: 947 (3.3%)

Dew Point:
  Unique values: 64
  Top 5 categories:
    33: 1,349 (4.7%)
    34: 1,156 (4.0%)
    30: 1,132 (3.9%)
    35: 1,124 (3.9%)
    32: 1,116 (3.9%)

Wind:
  Unique values: 18
  Top 5 categories:
    W: 3,273 (11.4%)
    NW: 3,173 (11.0%)
    NNW: 2,570 (8.9%)
    WSW: 2,567 (8.9%)
    N: 2,453 (8.5%)

Condition

## Research Paper Dataset Section Summary

Based on the analysis above, here's the formatted information for your research paper's dataset section:

---

### Dataset

**Dataset Name and Source:** M1_final.csv (Flight Operations and Weather Dataset)

**Number of Samples:** 28,820 flight records

**Number of Features:** 23 features

**Feature Categories:**
- **Temporal Features (8):** MONTH, DAY_OF_MONTH, DAY_OF_WEEK, CRS_DEP_M, DEP_TIME_M, CRS_ARR_M, sch_dep, sch_arr
- **Flight Operational Features (5):** OP_UNIQUE_CARRIER, TAIL_NUM, DEST, CRS_ELAPSED_TIME, DISTANCE  
- **Weather Features (8):** Temperature, Dew Point, Humidity, Wind, Wind Speed, Wind Gust, Pressure, Condition
- **Target Variables (2):** DEP_DELAY, TAXI_OUT

**Target Variable:** DEP_DELAY (Departure delay in minutes)
- Range: -22 to 1,276 minutes
- Mean: 6.37 minutes
- Standard Deviation: 38.74 minutes
- Negative values indicate early departures, 0 indicates on-time, positive values indicate delays

**Class Distribution (Delay Categories):**
- Early/On-time departures (≤0 min): 21,015 flights (72.9%)
- Short delays (1-15 min): 3,931 flights (13.6%)
- Medium delays (16-60 min): 2,556 flights (8.9%)
- Long delays (>60 min): 1,318 flights (4.6%)

**Missing/Duplicate Records:** 
- Missing values: 2 records (0.01%) in Wind feature only
- Duplicate records: 0
- Data completeness: 99.99%

**Time Period:** Data spans 3 months (January, November, December) with comprehensive coverage across all days of the week

**Data Quality:** High-quality dataset with minimal missing data and no duplicates

**Sampling Characteristics:** 
- 9 unique airline carriers
- 65 destination airports  
- 25 distinct weather conditions
- 2,092 unique aircraft tail numbers

---

In [13]:
# Final Summary Statistics for Research Paper
print("="*60)
print("FINAL DATASET SUMMARY FOR RESEARCH PAPER")
print("="*60)

summary_stats = {
    'Dataset Name': 'M1_final.csv (Flight Operations and Weather Dataset)',
    'Total Samples': f"{data.shape[0]:,}",
    'Total Features': f"{data.shape[1]:,}",
    'Target Variable': 'DEP_DELAY (Departure delay in minutes)',
    'Data Completeness': f"{((data.count().sum()) / (data.shape[0] * data.shape[1]) * 100):.2f}%",
    'Missing Values': f"{data.isnull().sum().sum():,} ({(data.isnull().sum().sum()/data.size)*100:.3f}%)",
    'Duplicate Records': f"{data.duplicated().sum():,}",
    'Memory Usage': f"{data.memory_usage(deep=True).sum() / 1024**2:.2f} MB",
    'Time Period': f"{data['MONTH'].nunique()} months (Jan, Nov, Dec)",
    'Airlines': f"{data['OP_UNIQUE_CARRIER'].nunique()} unique carriers",
    'Destinations': f"{data['DEST'].nunique()} airports",
    'Weather Conditions': f"{data['Condition'].nunique()} different conditions"
}

for key, value in summary_stats.items():
    print(f"{key:<20}: {value}")

print(f"\nDelay Distribution Summary:")
print(f"{'Category':<15} | {'Count':<10} | {'Percentage'}")
print("-" * 40)
delay_categories = pd.cut(data['DEP_DELAY'], 
                         bins=[-float('inf'), 0, 15, 60, float('inf')],
                         labels=['Early/OnTime', 'Short_Delay', 'Medium_Delay', 'Long_Delay'])
delay_dist = delay_categories.value_counts()
for category, count in delay_dist.items():
    percentage = (count / len(data)) * 100
    print(f"{category:<15} | {count:<10,} | {percentage:>6.1f}%")

print(f"\nThis analysis provides all required information for the dataset section of your research paper.")
print("="*60)

FINAL DATASET SUMMARY FOR RESEARCH PAPER
Dataset Name        : M1_final.csv (Flight Operations and Weather Dataset)
Total Samples       : 28,820
Total Features      : 23
Target Variable     : DEP_DELAY (Departure delay in minutes)
Data Completeness   : 100.00%
Missing Values      : 2 (0.000%)
Duplicate Records   : 0
Memory Usage        : 12.56 MB
Time Period         : 3 months (Jan, Nov, Dec)
Airlines            : 9 unique carriers
Destinations        : 65 airports
Weather Conditions  : 25 different conditions

Delay Distribution Summary:
Category        | Count      | Percentage
----------------------------------------
Early/OnTime    | 21,015     |   72.9%
Short_Delay     | 3,931      |   13.6%
Medium_Delay    | 2,556      |    8.9%
Long_Delay      | 1,318      |    4.6%

This analysis provides all required information for the dataset section of your research paper.
